src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -1936,25 +1932,19 @@ void InstanceKlass::print_nonstatic_fields(FieldClosure* cl) {
     super->print_nonstatic_fields(cl);
   }
   ResourceMark rm;
-  fieldDescriptor fd;
   // In DebugInfo nonstatic fields are sorted by offset.
-  GrowableArray<Pair<int,int> > fields_sorted;
-  int i = 0;
+  GrowableArray<FieldInfo> fields_sorted;
   for (AllFieldStream fs(this); !fs.done(); fs.next()) {
     if (!fs.access_flags().is_static()) {
-      fd = fs.field_descriptor();
-      Pair<int,int> f(fs.offset(), fs.index());
-      fields_sorted.push(f);
-      i++;
+      fields_sorted.push(fs.to_FieldInfo());
     }
   }
-  if (i > 0) {
-    int length = i;
-    assert(length == fields_sorted.length(), "duh");
+  if (fields_sorted.length() > 0) {
     fields_sorted.sort(compare_fields_by_offset);
-    for (int i = 0; i < length; i++) {
-      fd.reinitialize(this, fields_sorted.at(i).second);
-      assert(!fd.is_static() && fd.offset() == fields_sorted.at(i).first, "only nonstatic fields");
+    fieldDescriptor fd;
+    for (auto it = fields_sorted.begin(); it != fields_sorted.end(); ++it) {
```

Comment:
Ah, that's now what I meant :) There is no need to use iterators here, just say pull `length` out of `fields_stored.length()`, and use the same old indexed loop:

```
 int length = fields_stored.length();
 if (length > 0) {
    fields_sorted.sort(compare_fields_by_offset);
    for (int i = 0; i < length; i++) {
      ...
```

Reply:
I know you haven't asked specifically for iterators, but to me this seems 'the' cleaner way, rather than plain old indexed loop. Or do you have doubts about this being inlined to effectively the same code?

Reply:
I think using an iterator is fine, that's up to you. I am a bit bothered by having this change also slip in a change to the `GrowableArray` API, however. Anyone who wants to review that part of the code will miss that change.

I'd like to suggest that you change to an indexed for loop, and you make a follow-up PR with your GA changes and where you switch the loop.